shared/middleware/files_api.rb

```diff
@@ -487,10 +487,10 @@ def files_put_file(encrypted_channel_id, filename, body)
     end
     new_entry_hash = JSON.parse new_entry_json
     # Replace downcased filename with original filename (to preserve case in the manifest)
-    new_entry_hash['filename'] = filename
+    new_entry_hash['filename'] = CGI.unescape(filename)
     manifest_is_unchanged = false
 
-    existing_entry = manifest.detect {|e| e['filename'].downcase == filename.downcase}
+    existing_entry = manifest.detect {|e| e['filename'].downcase == CGI.unescape(filename).downcase}
```

Comment:
Nit: Probably not _very_ expensive, but since we're always going to call `CGI.unescape` in this method we might as well only call it once and reuse the result - then we won't be doing it in a loop here.  Likewise with the `downcase` variant which shows up in a couple of loops.